irc3/plugins/slack.py

```diff
@@ -0,0 +1,247 @@
+        self.channels.pop('#', None)
+        self.channels.pop('hash', None)
+
+    async def api_call(self, method, data=None):
+        """Slack API call."""
+        async with aiohttp.ClientSession() as session:
+            form = aiohttp.FormData(data or {})
+            form.add_field('token', self.config['token'])
+            async with session.post(
+                'https://slack.com/api/{0}'.format(method), data=form
+                    ) as response:
+                if response.status != 200:
+                    raise Exception(
+                        '{0} with {1} failed.'.format(method, data)
+                    )
+                return await response.json()
+
+    def server_ready(self):
+        irc3.asyncio.ensure_future(self.connect())
+        self.bot.log.debug('Listening to Slack')
+
+    def parse_text(self, message):
+
+        def getChannelById(matchobj):
+            return matchobj.group('readable') or '#{0}'.format(
+                self.slack_channels[matchobj.group('channelId')]['name']
+            )
+
+        def getUserById(matchobj):
+            return matchobj.group('readable') or '@{0}'.format(
+                self.slack_users[matchobj.group('userId')]
+            )
+
+        def getEmoji(matchobj):
+            emoji = matchobj.group('emoji')
+            return EMOJIS.get(emoji, emoji)
+
+        matches = [
+            (r'\n|\r\n|\r', ''),
+            (r'<!channel>', '@channel'),
+            (r'<!group>', '@group'),
+            (r'<!everyone>', '@everyone'),
+            (r'<#(?P<channelId>C\w+)\|?(?P<readable>\w+)?>', getChannelById),
+            (r'<@(?P<userId>U\w+)\|?(?P<readable>\w+)?>', getUserById),
+            (r':(?P<emoji>\w+):', getEmoji),
+            (r'<.+?\|(.+?)>', r'\g<0>'),
+            (r'&lt', '<'),
+            (r'&gt', '>'),
+            (r'&amp', '&'),
+        ]
+        for match in matches:
+            message = re.sub(*match, string=message)
+        return message
+
+    async def connect(self):
+        channels = await self.api_call('channels.list')
+        groups = await self.api_call('groups.list')
+        for channel in channels.get('channels', []):
+            self.slack_channels[channel['id']] = channel
+            if channel['name'] in self.channels:
+                self.channels[channel['id']] = self.channels.pop(
+                    channel['name']
+                )
+        for channel in groups.get('groups', []):
+            self.slack_channels[channel['id']] = channel
+            if channel['name'] in self.channels:
+                self.channels[channel['id']] = self.channels.pop(
+                    channel['name']
+                )
+        users = await self.api_call('users.list')
+        for user in users['members']:
+            self.slack_users[user['id']] = user['name']
+        rtm = await self.api_call('rtm.start')
+        if not rtm['ok']:
+            raise Exception('Error connecting to RTM')
+        while True:
```

Comment:
What happens if an exception occurs in this loop ?

Reply:
hrm, that is a fair point, i think that the with statements will catch the exceptions, but i will double check and handle them.